### PR TITLE
274 imm1 plugin change

### LINKED ISSUE #274: only use imm1 plugin now
For all detectors, use just the `IMM1:` plugin and remove immout, imm0, and imm2

### PR DESCRIPTION
**Pull request by qzhang234**
- FIX #274

### CODE REVIEW

**Comment by prjemian** [comment]
There's a lot of old code that is commented out.  Can these lines be deleted?  The code would be much easier to examine.

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 1]
```diff
@@ -23,16 +23,16 @@
 import struct
 
 
-class IMMnLocal(Device):
-    """
-    local interface to the IMM1, & IMM2 plugins
-    """
+# class IMMnLocal(Device):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:26]: remove this commented code

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 2]
```diff
@@ -23,16 +23,16 @@
...
-    )
-    num_captured = Component(EpicsSignalRO, "NumCaptured_RBV")
+#     capture = Component(EpicsSignalWithRBV, "Capture", kind="config")
+#     file_format = Component(
+#         EpicsSignalWithRBV, "NDFileIMM_format", string=True, kind="config"
+#     )
+#     num_captured = Component(EpicsSignalRO, "NumCaptured_RBV")
 
 
 class IMMoutLocal(Device):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:38]: Since all the IMM-specific PVs are not used by this class, looks as if it could be replaced completely by [`ophyd.areadetector.FilePlugin_V34`](https://github.com/bluesky/ophyd/blob/dd4b3e389a0202ecacce39fc3965d703c616b0d4/ophyd/areadetector/plugins.py#L1095)

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 3]
```diff
@@ -51,37 +51,39 @@ class IMMoutLocal(Device):
...
 class IMM_DeviceMixinBase(Device):
     """
     attributes used by area detectors using the IMM Plugin
     """
-
-    immout = Component(IMMoutLocal, "IMMout:")
-    imm1 = Component(IMMnLocal, "IMM1:")
-    imm2 = Component(IMMnLocal, "IMM2:")
+    # imm1 = Component(IMMnLocal, "IMM1:")
+    imm1 = Component(IMMoutLocal, "IMM1:")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:79]: replace `IMMoutLocal` with `phyd.areadetector.FilePlugin_V34`

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 4]
```diff
@@ -51,37 +51,39 @@ class IMMoutLocal(Device):
...
 
     def setIMM_Cmprs(self):
         """
         Set all IMM plugins for compression.
         """
         # from SPEC macro: ccdset_compr_params_ad_Lambda
-        for plugin in (self.imm1, self.imm2, self.immout):
+        for plugin in (self.imm1):
             if plugin.file_format.get() not in (1, "IMM_Cmprs"):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:87]: there is no `file_format` attribute now - what to do here?

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 5]
```diff
@@ -95,7 +97,7 @@ def setIMM_Raw(self):
         Set all IMM plugins for raw (uncompressed).
         """
         # from SPEC macro: ccdset_RawMode_params_ad_Lambda
-        for plugin in (self.imm1, self.imm2, self.immout):
+        for plugin in (self.imm1):
             if plugin.file_format.get() not in (0, "IMM_Raw"):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:101]: same as above, no `file_format` attribute now - what to do here?

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 6]
```diff
@@ -314,7 +314,7 @@ def describe(self):
         return res
 
 
-class Lambda750kLocal(AD_AcquireDetectorBase, IMM_DeviceMixinBase, DM_DeviceMixinAreaDetector, Device):
+class Lambda750kLocal(IMM_DeviceMixinBase, DM_DeviceMixinAreaDetector, AD_AcquireDetectorBase):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:317]: good

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 7]
```diff
@@ -477,8 +485,9 @@ def watch_acquire(value, old_value, **kwargs):
         time.sleep(0.005)  # wait for the shutter to move out of the way
         self.cam.state.subscribe(watch_state)
         self.imm1.capture.subscribe(watch_acquire)
-        for plugin in (self.imm0, self.imm1, self.imm2):
-            plugin.capture.put(1, wait=False)
+        # for plugin in (self.imm0, self.imm1, self.imm2):
+        #     plugin.capture.put(1, wait=False)
+        self.imm1.capture.put(1, wait=False)
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:490]: Is this line a duplicate of the one below?  Can it be removed?  Otherwise comment why it is needed twice.

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 8]
```diff
@@ -386,8 +390,8 @@ def staging_setup_DM(self, *args, **kwargs):
         acquire_period = args[4]
         # logger.debug(f"staging_setup_DM({args})")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:392]: remove these two lines

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 9]
```diff
@@ -398,16 +402,19 @@ def staging_setup_DM(self, *args, **kwargs):
         self.imm1.stage_sigs["file_name"] = self._file_name
         self.imm1.stage_sigs["num_capture"] = num_images
         self.imm1.stage_sigs["file_number"] = 1
-        self.imm1.stage_sigs["file_format"] = "IMM_Cmprs"
+        # self.imm1.stage_sigs["file_format"] = "IMM_Cmprs"
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:405]: remove these two lines

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 10]
```diff
@@ -398,16 +402,19 @@ def staging_setup_DM(self, *args, **kwargs):
...
 
     def stage(self):
         super().stage()
-        root = os.path.join("/", "home", "8-id-i/")
-        if self._file_path.startswith("/data/"):
-            self._file_path = self._file_path[len("/data/") :]
-        elif self._file_path.startswith("/home/8-id-i/"):
-            self._file_path = self._file_path[len("/home/8-id-i/") :]
+        root = os.path.join("/", "home", "8ididata/")
+        # print(root)
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:412]: remove these commented lines

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 11]
```diff
@@ -477,8 +485,9 @@ def watch_acquire(value, old_value, **kwargs):
         time.sleep(0.005)  # wait for the shutter to move out of the way
         self.cam.state.subscribe(watch_state)
         self.imm1.capture.subscribe(watch_acquire)
-        for plugin in (self.imm0, self.imm1, self.imm2):
-            plugin.capture.put(1, wait=False)
+        # for plugin in (self.imm0, self.imm1, self.imm2):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:488]: remove these two lines

File: profile_bluesky/startup/instrument/devices/shutters.py [hunk 12]
```diff
@@ -34,7 +42,8 @@
     # suspend_APS_current = SuspendFloor(aps.current, 2, resume_thresh=10, sleep=100)
     # RE.install_suspender(suspend_APS_current)
 
-    shutter = EpicsOnOffShutter("8idi:Unidig1Bo13", name="shutter")
+    # shutter = EpicsOnOffShutter("8idi:Unidig1Bo13", name="shutter")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/shutters.py:45]: just don't forget to switch back to the real shutter when needed

File: profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
+
+
+
+"""
+Acquire an XPCS measurement with a supported area detector
+"""
+
+__all__ = """
+    AD_Acquire
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py:9]: change to `Detector_Acq`

Comment (prjemian) [reply profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py:9]: otherwise, this file should NOT be changed before this PR is merged, even though the `preconfigure()` method is not complete

File: profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    def preconfigure(self,
+                trigger_mode,
+               file_name,
+               acquire_time,
+               acquire_period,
+               num_images,
+               path=None):   # Each detector needs this method. Copy into the detector classes.
+        """
+        Setup staging, trigger mode
+        """
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py:67]: This method needs to be completed but that work is not related to this PR.

**Review by prjemian — CHANGES_REQUESTED** [review]
changes as noted

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 1 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:26]: Done

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 2 above)
Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:38]: We switched to `ophyd.areadetector.FilePlugin_V34` as a base class for the Imm plugin but that still needs testing.

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 3 above)
Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:79]: This does not need a change with the revision above.

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 4 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:87]: To self: Check with Suresh whether `setIMM_Cmprs` and `setIMM_Raw` are actually being used

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 5 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:101]: Same as above

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 2 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:38]: Source has been changed, commit is pending

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 1 above)
Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:27]: instead, leave it commented since it is going away with this PR

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py (hunk 4 above)
Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:84]: In the code, only appears here (so, not in use anywhere else in bluesky).

Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:84]: comment out now, consider to delete later (after this PR is merged)

File: profile_bluesky/startup/instrument/devices/lambda_750k.py (hunk 8 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/lambda_750k.py:392]: done

File: profile_bluesky/startup/instrument/devices/lambda_750k.py (hunk 9 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/lambda_750k.py:405]: done

File: profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py (hunk 13 above)
Comment (qzhang234) [reply profile_bluesky/startup/instrument/plans/xpcs_acquire_2.py:9]: This file is a *sketch* of code now, not actually used.  It is unrelated to the PR except for when we thought about it.  Got to write our thoughts somewhere.

**Review by prjemian — APPROVED** [review]
Good enough, let's merge this thing to branch `QZ_Test`

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 15]
```diff
@@ -370,6 +373,7 @@ def plugin_file_name(self):
         """
         # cut the path from file name
         return os.path.basename(self.imm1.full_file_name.get())
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:375]: refers to imm1 plugin here, must change per this PR's intent

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 16]
```diff
@@ -386,9 +390,6 @@ def staging_setup_DM(self, *args, **kwargs):
         acquire_period = args[4]
         # logger.debug(f"staging_setup_DM({args})")
 
-        if self._file_path.startswith("/home/8-id-i/"):
-            self._file_path = "/data/" + self._file_path.lstrip("/home/8-id-i/")
-
         self.cam.stage_sigs["num_images"] = num_images
         # replaced by: self.cam.setTime(acquire_time, acquire_period)
         self.imm1.stage_sigs["enable"] = 1
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:395]: imm1 plugin staging here needs to be changed

**Review by prjemian — APPROVED** [review]
LGTM
